src/rules/no_misused_new.rs

```diff
@@ -0,0 +1,212 @@
+    visitor.visit_module(&module, &module);
+  }
+
+  fn code(&self) -> &'static str {
+    "no-misused-new"
+  }
+}
+
+pub struct NoMisusedNewVisitor {
+  context: Context,
+}
+
+impl NoMisusedNewVisitor {
+  fn new(context: Context) -> Self {
+    Self { context }
+  }
+
+  fn match_parent_type(&self, parent: &Ident, return_type: &TsTypeAnn) -> bool {
+    if let TsType::TsTypeRef(type_ref) = &*return_type.type_ann {
+      if let TsEntityName::Ident(ident) = &type_ref.type_name {
+        return ident.sym == parent.sym;
+      }
+    }
+
+    false
+  }
+
+  fn is_constructor_keyword(&self, ident: &Ident) -> bool {
+    JsWord::from("constructor") == ident.sym
+  }
+}
+
+impl Visit for NoMisusedNewVisitor {
+  fn visit_ts_type_alias_decl(
+    &mut self,
+    t: &TsTypeAliasDecl,
+    _parent: &dyn Node,
+  ) {
+    if let TsType::TsTypeLit(lit) = &*t.type_ann {
+      for member in &lit.members {
+        if let TsMethodSignature(signature) = &member {
+          if let Expr::Ident(ident) = &*signature.key {
+            if self.is_constructor_keyword(&ident) {
+              self.context.add_diagnostic(
+                ident.span,
+                "no-misused-new",
+                "Type aliases cannot be constructed, only classes",
+              );
+            }
+          }
+        }
+      }
+    }
+  }
+
+  fn visit_ts_interface_decl(
+    &mut self,
+    n: &TsInterfaceDecl,
+    _parent: &dyn Node,
+  ) {
+    for member in &n.body.body {
+      match &member {
+        TsMethodSignature(signature) => {
+          if let Expr::Ident(ident) = &*signature.key {
+            if self.is_constructor_keyword(&ident) {
+              // constructor
+              self.context.add_diagnostic(
+                signature.span,
+                "no-misused-new",
+                "Interfaces cannot be constructed, only classes",
+              );
+            }
+          }
+        }
+        TsConstructSignatureDecl(signature) => {
+          if signature.type_ann.is_some()
+            && self
+              .match_parent_type(&n.id, &signature.type_ann.as_ref().unwrap())
+          {
+            self.context.add_diagnostic(
+              signature.span,
+              "no-misused-new",
+              "Interfaces cannot be constructed, only classes",
+            );
+          }
+        }
+        _ => {}
+      }
+    }
+  }
+
+  fn visit_class_decl(&mut self, expr: &ClassDecl, _parent: &dyn Node) {
+    for member in &expr.class.body {
+      if let ClassMember::Method(method) = member {
+        if method.function.return_type.is_some()
+          && self.match_parent_type(
+            &expr.ident,
+            &method.function.return_type.as_ref().unwrap(),
+          )
+        {
+          // new
+          self.context.add_diagnostic(
+            method.span,
+            "no-misused-new",
+            "Class cannot have method named `new`.",
+          );
+        }
+      }
+    }
+  }
```

Comment:
There should probably be call to `swc_ecma_visit::visit_class_decl(self, expr, parent)` otherwise nested classes won't get inspected